x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -368,14 +367,10 @@ private void createOAuth2Tokens(
                 traceLog("create token", new IllegalArgumentException("originating client authentication must be provided"))
             );
         } else {
-            final Authentication tokenAuth = new Authentication(
-                authentication.getUser(),
-                authentication.getAuthenticatedBy(),
-                authentication.getLookedUpBy(),
-                tokenVersion,
-                AuthenticationType.TOKEN,
-                authentication.getMetadata()
-            );
+            final Authentication tokenAuth = authentication.token().maybeRewriteForOlderVersion(tokenVersion);
+            if (authentication.isAssignedToDomain() && false == tokenAuth.isAssignedToDomain()) {
+                logger.info("Rewriting authentication [" + authentication + "] for token without domain");
+            }
```

Comment:
I don't think we should rewrite here. But this is a good point. The Authentication object now has two places that handles BWC write (`writeTo` and `maybeRewriteForOlderVersion`) and I think we should consolidate them so the caller should not worry about it. I suggest we do something like:

```
class Authentication {
  public void writeTo(StreamOutput out) {
    if (out.getVersion().before(getVersion()) {
      maybeRewriteForOlderVersion(out.getVersion).writeTo(out);
    } else {
      // Rest of the code
    }
  }
}
```
So when UserToken calls `Authentication#writeTo` for serialisation, it can be sure all BWC is in place without explicity asking for it. This is also helpful in all other places where we need serialise authentication. I wonder whether this could already be a bug.

Reply:
We discussed, the rewrite is suitable here, for pragmatic/minimalistic reasons, but we need to abstract away the downgrade to the serialization part.